### PR TITLE
feat(crafters): add SlidingWindowSegmenter for nlp

### PR DESCRIPTION
**Pull request by fhaase2**
I use this for pretty long text documents. Might be useful for others as well?

### CODE REVIEW

**Comment by hanxiao** [comment]
hi @fhaase2 thanks for your contribution ❤️ . One remark when you do the commit. Can you properly set `git config` user and email and match it with your github account (@fhaase2)? This is important for:
- Your credit can be counted correctly.
- Our CLA bot can work properly.

Thanks a lot!

Right now you see none of them match to real account

image

**Comment by fhaase2** [comment]
hi @hanxiao, configured it correctly now.

**Comment by hanxiao** [comment]
recheckcla

File: jina/executors/crafters/nlp/split.py [hunk 1]
```diff
@@ -97,3 +99,73 @@ def craft(self, buffer: bytes, doc_id: int, *args, **kwargs) -> List[Dict]:
...
+            
+        results = []
+        for idx, s in enumerate(chunks):
+            if self.min_substring_len <= len(s):
+                results.append(dict(
+                    doc_id=doc_id,
+                    text=s,
+                    offset=idx,
+                    weight=1.0,
+                    length=len(chunks)))
```
Comment (hanxiao) [inline jina/executors/crafters/nlp/split.py:170]:
```suggestion
                    text=s,
                    offset=idx,
                    weight=1.0))
```

File: jina/executors/crafters/nlp/split.py [hunk 2]
```diff
@@ -97,3 +99,73 @@ def craft(self, buffer: bytes, doc_id: int, *args, **kwargs) -> List[Dict]:
...
+        if self.window_size <= 0:
+            self.logger.warning(
+                'the window_size (={}) should be larger than zero'.format(
+                    self.window_size))
+        if self.step_size > self.window_size:
+            self.logger.warning(
+                'the step_size (={}) should not be larger than the window_size (={})'.format(
+                    self.window_size, self.step_size))
+
+    def craft(self, buffer: bytes, doc_id: int, *args, **kwargs) -> List[Dict]:
```
Comment (hanxiao) [inline jina/executors/crafters/nlp/split.py:134]:
```suggestion
    def craft(self, text: str, doc_id: int, *args, **kwargs) -> List[Dict]:
```

File: jina/executors/crafters/nlp/split.py [hunk 3]
```diff
@@ -97,3 +99,73 @@ def craft(self, buffer: bytes, doc_id: int, *args, **kwargs) -> List[Dict]:
...
+            while True:
+                yield iter(d)
+                try:
+                    d.append(next(i))
+                except StopIteration:
+                    return
+                d.extend(next(i, None)
+                         for _ in range(step-1))
+
+        text = buffer.decode('utf-8')
```
Comment (hanxiao) [inline jina/executors/crafters/nlp/split.py:158]: starting from `v0.2.0` no need on `buffer.decode()`, you can directly use `text`, see https://docs.jina.ai/chapters/proto/docs.html#document 
```suggestion
```

File: jina/executors/crafters/nlp/split.py [hunk 4]
```diff
@@ -97,3 +99,73 @@ def craft(self, buffer: bytes, doc_id: int, *args, **kwargs) -> List[Dict]:
...
+        chunks = ["".join(filter(None, list(chunk))) for chunk in sliding_window(
```
Comment (hanxiao) [inline jina/executors/crafters/nlp/split.py:159]: single quote is the preferred coding style of this project

```suggestion
        chunks = [''.join(filter(None, list(chunk))) for chunk in sliding_window(
```

File: tests/executors/crafters/nlp/split.py [hunk 5]
```diff
@@ -55,6 +55,16 @@ def test_jieba_crafter(self):
         crafted_chunk_list = jieba_crafter.craft(buffer, 0)
         self.assertEqual(len(crafted_chunk_list), 14)
 
+    def test_sliding_window_segmenter(self):
+        window_size = 20
+        step_size = 10
+        sliding_window_segmenter = SlidingWindowSegmenter(
+            window_size=window_size, step_size=step_size)
+        buffer = b'It is a sunny day!!!! When Andy comes back, we are going to the zoo.'
+        crafted_chunk_list = sliding_window_segmenter.craft(buffer, 0)
```
Comment (hanxiao) [inline tests/executors/crafters/nlp/split.py:64]:
```suggestion
        text = 'It is a sunny day!!!! When Andy comes back, we are going to the zoo.'
        crafted_chunk_list = sliding_window_segmenter.craft(text, 0)
```

**Comment by fhaase2** [comment]
Thanks for your suggestions. Updated it.

**Pull request merged**